packages/eslint-plugin/src/rules/require-object-type-annotations.ts

```diff
@@ -0,0 +1,55 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import type { RuleListener } from '@typescript-eslint/utils/src/ts-eslint';
+import type * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'require-object-type-annotations',
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: '',
+      requiresTypeChecking: true,
+      recommended: false,
+    },
+    messages: {
+      forbidden: 'Object is missing type annotation.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  create: (context): RuleListener => {
+    const listener = (esNode: TSESTree.ObjectExpression): void => {
+      const parserServices = util.getParserServices(context);
+      const checker = parserServices.program.getTypeChecker();
+
+      const tsNode: ts.ObjectLiteralExpression =
+        parserServices.esTreeNodeToTSNodeMap.get(esNode);
+
+      const type = checker.getTypeAtLocation(tsNode);
+
+      // Allow empty objects
+      if (type.getProperties().length === 0) {
+        return;
+      }
+
+      const contextualType = checker.getContextualType(tsNode);
+
+      if (
+        contextualType === undefined ||
+        // Needed to catch object passed as a function argument where the parameter type is generic,
+        // e.g. an identity function
+        contextualType?.getSymbol()?.name === '__object'
```

Comment:
I think you could do this with
```
(contextualType.flags & ts.TypeFlags.Intrinsic) === 0 &&
checker.typeToString(contextualType) === 'object'
```

Which is a little less abstract than using the `__` string

Reply:
Interesting, thanks! I will have to try that and see if my tests still pass.

It seems like `Intrinsic` doesn't exist? Also, what does "intrinsic" mean exactly?

<img width="733" alt="image" src="https://user-images.githubusercontent.com/921609/190866125-2893ba4e-b064-448d-a3c3-1a7636207cc3.png">